tools/Azure.Mcp.Tools.EventGrid/src/Services/EventGridService.cs

```diff
@@ -82,6 +83,201 @@ public async Task<List<EventGridSubscriptionInfo>> GetSubscriptionsAsync(
+            operationId, topicName, resourceGroup, subscription);
+
+        try
+        {
+            var subscriptionResource = await _subscriptionService.GetSubscription(subscription, tenant, retryPolicy);
+
+            // Find the topic to get its endpoint and access key
+            var topic = await FindTopic(subscriptionResource, resourceGroup, topicName);
+            if (topic == null)
+            {
+                var errorMessage = $"Event Grid topic '{topicName}' not found in resource group '{resourceGroup}'. Make sure the topic exists and you have access to it.";
+                _logger.LogError(errorMessage);
+                return new EventPublishResult(
+                    Status: "Failed",
+                    Message: errorMessage,
+                    PublishedEventCount: 0,
+                    OperationId: operationId,
+                    PublishedAt: DateTime.UtcNow);
+            }
+
+            if (topic.Data.Endpoint == null)
+            {
+                var errorMessage = $"Event Grid topic '{topicName}' does not have a valid endpoint.";
+                _logger.LogError(errorMessage);
+                return new EventPublishResult(
+                    Status: "Failed",
+                    Message: errorMessage,
+                    PublishedEventCount: 0,
+                    OperationId: operationId,
+                    PublishedAt: DateTime.UtcNow);
+            }
+
+            // Get the access key for the topic
+            var keys = await topic.GetSharedAccessKeysAsync();
+            var accessKey = keys.Value.Key1;
+
+            if (string.IsNullOrEmpty(accessKey))
+            {
+                var errorMessage = $"Unable to retrieve access key for Event Grid topic '{topicName}'.";
+                _logger.LogError(errorMessage);
+                return new EventPublishResult(
+                    Status: "Failed",
+                    Message: errorMessage,
+                    PublishedEventCount: 0,
+                    OperationId: operationId,
+                    PublishedAt: DateTime.UtcNow);
+            }
+
+            // Parse and validate event data
+            var events = ParseAndValidateEventData(eventData, eventSchema ?? "EventGridEvent");
+
+            // Create EventGridPublisherClient and publish events
+            var credential = new AzureKeyCredential(accessKey);
+            var client = new EventGridPublisherClient(topic.Data.Endpoint, credential);
+
+            // Convert to EventGridEvent objects for publishing
+            var eventGridEvents = events.Select(eventData => new EventGridEvent(
+                subject: eventData.Subject,
+                eventType: eventData.EventType,
+                dataVersion: eventData.DataVersion,
+                data: BinaryData.FromString(eventData.Data ?? "{}")
+            )
+            {
+                Id = eventData.Id,
+                EventTime = eventData.EventTime
+            }).ToList();
```

Comment:
The client takes IEnumerable<T>, you don't have to realise the IEnumerable into a List, so it can be lazily evaluated.